### PR TITLE
Allow coordgen to run without maeparser.

### PR DESCRIPTION
**Pull request by d-b-w**
See internal Schrödinger rationale here:
https://jira.schrodinger.com/browse/CRDGEN-252

Set the CMake define COORDGEN_BUNDLE_TEMPLATES to include the
coordgen templates directly in the library. When the templates
are bundled, maeparser is no longer used, so it isn't linked
into the library, either.
    
This is useful for a couple of reasons:

- We definitely see bugs in template discovery
- 3rd party packagers seem to have trouble with building maeparser.
 - Allow use in other novel places that can't use the boost:: infrastructure required by maeparser.
    
maeparser is always required for running the unit tests.
    
Includes a short Python script for generating the C++ files
describing the templates. This Python script currently requires
use of the Schrödinger Python tools, so it will need to be
manually updated if the templates are updated. There is a test
(https://github.com/schrodinger/coordgenlibs/pull/53) that will fail if a new template is added but isn't added to
the compiled templates.

**Note to reviewers**
One of the commits here is entirely autogenerated code. I'd recommend avoiding looking at it because it is very long and duplicative, unless it is helpful in reading the Python file that is used to generate the code.

### CODE REVIEW

File: mol_generator.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    static const auto t = create_templates();
+    return t;
+}}
+}}
+
+"""
+
+header_path = os.path.join(template_dir, f'{NAME}.h')
+with open(header_path, 'w') as fh:
+    fh.write(header)
```
Comment (ricrogz) [inline mol_generator.py]: I think you need to format `header`with the number of templates.

File: mol_generator.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+implementation_path = os.path.join(template_dir, f'{NAME}.cpp')
+with open(implementation_path, 'w') as fh:
+    fh.write(implementation)
```
Comment (ricrogz) [inline mol_generator.py]: Format with `NAME` and `body`.

File: mol_generator.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    for b in st.bond:
+        a1 = b.atom1.index - 1
+        a2 = b.atom2.index - 1
+        t.append(_bond.format(a1=a1, a2=a2, order=b.order))
+    t.append('molecules[i] = molecule;')
+    t.append('++i;')
+    t = '\n'.join(t)
+    t = textwrap.indent(t, '    ')
+
+    return '\n{\n' + t + '\n}\n'
```
Comment (ricrogz) [inline mol_generator.py]: This is not being used, right?

File: mol_generator.py (hunk 2 above)
Comment (d-b-w) [reply mol_generator.py]: implementation is an fstring.

File: mol_generator.py (hunk 1 above)
Comment (d-b-w) [reply mol_generator.py]: It's an fstring, so it gets those from the scope.

File: mol_generator.py (hunk 2 above)
Comment (d-b-w) [reply mol_generator.py]: but this tells me that these are unclear (of course). I'll adjust.

File: mol_generator.py (hunk 3 above)
Comment (d-b-w) [reply mol_generator.py]: good call.

File: mol_generator.py (hunk 2 above)
Comment (d-b-w) [reply mol_generator.py]: refactored mol_generator.py

File: CMakeLists.txt [hunk 4]
```diff
@@ -44,25 +52,37 @@ if(COORDGEN_BUILD_SHARED_LIBS)
...
+
 # Dependencies
 if(TARGET maeparser)
     message(STATUS "Using externally defined maeparser target to "
     "build coordgen")
 else()
     include(CoordgenUtils)
-    set(MAEPARSER_BUILD_SHARED_LIBS ${COORDGEN_BUILD_SHARED_LIBS} CACHE BOOL "Library type for maeparser")
-    find_or_clone_maeparser()
+    if (USE_MAEPARSER)
```
[the remark this answers is not in the document]
Comment (d-b-w) [reply CMakeLists.txt]: I've merged these blocks

File: mol_generator.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+#pragma once
+
+{doc}
+
+#include <vector>
+
+class sketcherMinimizerMolecule;
+namespace schrodinger {{
+
+std::vector<sketcherMinimizerMolecule*> coordgen_templates();
```
Comment (greglandrum) [inline mol_generator.py]: I'd include a comment here indicating that the caller owns the pointers in the `std::vector` this returns.

File: mol_generator.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+# String that creates a molecule and adds it to a container
+# of sketcherMinimizerMolecule* at index i.
+#
+# C++ programmers: `{` and `}` are escaped by duplication, so
+# `{{` means `{`.
+_MOLECULE = """
+{{
+    auto molecule = new sketcherMinimizerMolecule();
+    std::array<std::tuple<int, float, float>, {atom_total}> atoms = {{{{
```
Comment (greglandrum) [inline mol_generator.py]: Given that this is just a template in a code-generation function, I don't think it's critical, but the duplication of the name `atoms` here makes my head hurt.
I'd rename this variable to `atomVect`

File: mol_generator.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+{atoms}
+    }}}};
+    std::array<std::array<int, 3>, {bond_total}> bonds = {{{{
```
Comment (greglandrum) [inline mol_generator.py]: I'd rename this variable to `bondVect`

File: sketcherMinimizer.cpp [hunk 8]
```diff
@@ -8,19 +8,25 @@
...
 
 #include "sketcherMinimizer.h"
-#include "sketcherMinimizerMaths.h"
 
+#include <algorithm>
+#include <cstdio>
+#include <queue>
+#include <stack>
+
+#ifdef BUNDLE_TEMPLATES
```
Comment (greglandrum) [inline sketcherMinimizer.cpp]: I think we should consider having two different options:
`BUNDLE_TEMPLATES` will bundle the default templates into the code so that you don't *need* to look for the template file in order to use coordgen.
`INCLUDE_MAEPARSER` controls whether or not the MAE parser is available. This could be useful in a future version where client-code can provide templates.

File: sketcherMinimizer.h [hunk 9]
```diff
@@ -467,7 +467,7 @@ class EXPORT_COORDGEN sketcherMinimizer
  * and bond orders will be ignored (i.e. no chiralities or stereo bonds will be
  * parsed).
  */
-EXPORT_COORDGEN sketcherMinimizerMolecule*
-mol_from_mae_block(schrodinger::mae::Block& block);
+// EXPORT_COORDGEN sketcherMinimizerMolecule*
```
Comment (greglandrum) [inline sketcherMinimizer.h]: might as well just delete these

File: CoordgenTemplates.h [hunk 10]
```diff
@@ -0,0 +1,23 @@
+
```
Comment (greglandrum) [inline CoordgenTemplates.h]: Does this file need to be added to the `install` list in CMakeLists.txt?
It looks like all the other header files are there, but I'm not sure all of them are actually needed (that'd be another PR to fix).

File: sketcherMaeReading.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
+#pragma once
```
Comment (greglandrum) [inline sketcherMaeReading.h]: Does this file need to be added to the `install` list in CMakeLists.txt?
It looks like all the other header files are there, but I'm not sure all of them are actually needed (that'd be another PR to fix).

File: CoordgenTemplates.h (hunk 10 above)
Comment (d-b-w) [reply CoordgenTemplates.h]: I totally didn't even think about that.  But I don't think so, because the function that CoordgenTemplates.h has hidden visibility, so a client can't use the function.

File: sketcherMaeReading.h (hunk 11 above)
Comment (d-b-w) [reply sketcherMaeReading.h]: I think that this is also private. We could expose it, though.

File: sketcherMinimizer.h (hunk 9 above)
Comment (d-b-w) [reply sketcherMinimizer.h]: thanks.

File: sketcherMinimizer.cpp (hunk 8 above)
Comment (d-b-w) [reply sketcherMinimizer.cpp]: Ricardo suggested _always_ bundling.

I think that dovetails with your suggestion, so that we'd do:

* always bundle the default templates
* if compiled with maeparser support, also allow override of template file.

So then the only define would be INCLUDE_MAEPARSER (on by default). What do you think of that?

Comment (greglandrum) [reply sketcherMinimizer.cpp]: Makes sense

File: sketcherMaeReading.h (hunk 11 above)
Comment (greglandrum) [reply sketcherMaeReading.h]: Na, I think it makes sense to do a separate PR to only install the bits that are actually required to use the library. 

I was just pointing it out here to be consistent; definitely not required

Comment (greglandrum) [reply sketcherMaeReading.h]: Na, I think it makes sense to do a separate PR to only install the bits that are actually required to use the library. 

I was just pointing it out here to be consistent; definitely not required

File: CMakeLists.txt [hunk 12]
```diff
@@ -118,10 +115,6 @@ install(FILES
     sketcherMinimizerStretchInteraction.h
     DESTINATION include/coordgen)
 
-install(FILES
-    templates.mae
-    DESTINATION share/coordgen)
-
```
Comment (ricrogz) [inline CMakeLists.txt]: We still need this if we enable `COORDGEN_USE_MAEPARSER`, right?

File: README.md [hunk 13]
```diff
@@ -11,14 +11,19 @@ Schrodinger intends to continue to contribute to this code as it still uses it i
 
 Examples and documentation will be added/improved over time
 
+## Templates
+
+Coordgen uses templates for some macrocycle systems. The default templates are in `templates.mae`. If you need additional
+local templates, use `CoordgenTemplates::setTemplateDir()`, and put them in a file called `user_templates.mae`.
```
Comment (ricrogz) [inline README.md]: This is now outdated, right?

We should mention that, by default, the templates are now hardcoded and that templates.mae does not get installed, unless we set `COORDGEN_USE_MAEPARSER`, which keeps the old behavior (install/use templates). Maybe also mentioned that the tests also use the templates file.

File: sketcherMinimizer.cpp [hunk 14]
```diff
@@ -3546,21 +3545,17 @@ void sketcherMinimizer::loadTemplates()
...
         normalizeTemplate(m);
     }
-#else
-    string filename = getTempFileProjDir() + "templates.mae";
-
-    loadTemplate(filename, m_templates.getTemplates());
+#ifdef USE_MAEPARSER
 
     filename = getUserTemplateFileName();
     loadTemplate(filename, m_templates.getTemplates());
```
Comment (ricrogz) [inline sketcherMinimizer.cpp]: Some silly thought for maybe another PR: if we move this a few lines up, before `m_templates.getTemplates() = coordgen_templates();`, do we get to overriding the default templates for free?

File: CMakeLists.txt (hunk 12 above)
Comment (d-b-w) [reply CMakeLists.txt]: no - the templates are always loaded from the library; "templates.mae" is never used. COORDGEN_USE_MAEPARSER allows override using "user_templates.mae".

Comment (ricrogz) [reply CMakeLists.txt]: Nevermind. GitHub's code reviews always get me confused. I can't get used to this formatting.

File: README.md (hunk 13 above)
Comment (ricrogz) [reply README.md]: I still think we should clarify that templates are now hardcoded, and `templates.mae` is provided mostly for testing.

Comment (d-b-w) [reply README.md]: I don't think this is outdated. The templates are in templates.mae. It's just that templates.mae is parsed to generate data that is in the library. If someone wanted to visualize the templates that we actually use, reading/displaying "templates.mae" would be the right choice. If someone wanted to update the templates that we use, editing "templates.mae" would be the right choice.

If this were internal, I'd build CoordgenTemplates.h and CoordgenTemplates.cpp at build time instead of at run time. I poke around at generating those files using maeparser.cpp at compile time, but that seems perhaps overly complex.

I'll add a comment about how to update templates.

File: sketcherMinimizer.cpp (hunk 14 above)
Comment (d-b-w) [reply sketcherMinimizer.cpp]: I didn't want to change user-visible behavior in this PR.

I agree, though, it might make sense for user templates to hit before the default templates.

**Comment by d-b-w** [comment]
Greg - I've been using this in my local RDKit and Schrodinger core suite, I think it's OK.
